File: multilive/platforms/kuaishou.py
"""快手直播（源自 kuaishou 方案，纯 HTTP，无浏览器）。

数据源 live_api/hot/list 与页面下拉加载一致：免登录免签名，
每页 50 个在播房间 + 4 档清晰度 CDN 直链（FLV，签名 24h 有效）。
m3u 语义：keep_stale=False——只保留「此刻在播」列表，下播即失效。
"""
import re
import time
import urllib.parse
from concurrent.futures import ThreadPoolExecutor, as_completed

from multilive.config import Source
from multilive.core import Room, Session, fmt_exc, log

NAME = 'kuaishou'
keep_stale = False

HOT_API = 'https://live.kuaishou.com/live_api/hot/list'
DEFAULT_PAGES = 50
MAX_PAGES = 50
PAGE_WORKERS = 5
MAX_SOURCE_WORKERS = 5
REQUEST_TIMEOUT = 20


def parse(line):
    t = line.strip()
    if not t:
        return []
    name = t
    pages = DEFAULT_PAGES
    if ':' in t.split('/')[-1]:
        name, _, pages_s = t.rpartition(':')
        try:
            pages = int(pages_s)
        except ValueError:
            pass
    name = (name.split('/live/')[-1].rstrip('/') or '').upper()
    if not re.fullmatch(r'[A-Z0-9]+', name or ''):
        return []
    src = Source(NAME, 'list', name)
    src.meta = pages
    return [src]


def fetch_page(sess, source, page):
    url = f'{HOT_API}?{urllib.parse.urlencode({
        "type": source, "filterType": 0, "page": page, "pageSize": 24})}'
    try:
        _st, j, _ = sess.get_json(
            url, timeout=REQUEST_TIMEOUT,
            referer=f'https://live.kuaishou.com/live/{source}')
        return page, (j.get('data') or {}).get('list') or []
    except Exception as e:
        log().info('  [%s] 第%d页失败: %s', source, page, fmt_exc(e))
        return page, []


def best_play_url(room):
    reps = []
    for pu in room.get('playUrls') or []:
        reps.extend((pu.get('adaptationSet') or {}).get('representation') or [])
    if not reps:
        return None
    reps = sorted(reps, key=lambda x: (x.get('level', 0), x.get('bitrate', 0)))
    return reps[-1].get('url')


def fetch_source(sess, source, pages):
    rooms = {}
    t0 = time.time()
    with ThreadPoolExecutor(max_workers=min(PAGE_WORKERS, pages)) as pool:
        futs = {pool.submit(fetch_page, sess, source, p): p
                for p in range(1, pages + 1)}
        for fut in as_completed(futs):
            _page, items = fut.result()
            for room in items:
                lid = room.get('id')
                if lid and lid not in rooms:
                    rooms[lid] = room
    log().info('  [%s] %d页抓完: %d 个不重复房间, 用时 %.1fs',
               source, pages, len(rooms), time.time() - t0)
    return rooms


def fetch(sources, ctx):
    log().info('[kuaishou] %d 个来源', len(sources))
    all_rooms = {}
    with ThreadPoolExecutor(max_workers=MAX_SOURCE_WORKERS) as pool:
        futs = {}
        for s in sources:
            sess = Session()
            pages = min(s.meta if isinstance(s.meta, int) else DEFAULT_PAGES,
                        MAX_PAGES)
            if ctx.pages_cap:
                pages = min(pages, ctx.pages_cap)
            futs[pool.submit(fetch_source, sess, s.target, pages)] = s
        for fut, s in futs.items():
            try:
                for lid, room in fut.result().items():
                    if lid not in all_rooms:
                        all_rooms[lid] = room
            except Exception as e:
                log().warning('[kuaishou] 来源 %s 失败: %s', s.target, fmt_exc(e))

    out = []
    for lid, room in all_rooms.items():
        author = room.get('author') or {}
        game = ((room.get('gameInfo') or {}).get('name') or '').strip() or NAME
        url = best_play_url(room)
        if not url:
            continue
        out.append(Room(platform=NAME, rid=str(lid),
                        title=(room.get('caption') or '').strip(),
                        nickname=(author.get('name') or '').strip(),
                        url=url.replace('http://', 'https://'),
                        group=game,
                        avatar=(room.get('cover') or '')))
    log().info('[kuaishou] 完成: %d 房间（无地址已跳过）', len(out))
    return out

File: internal/platforms/kuaishou/kuaishou.go
// Package kuaishou 快手直播：live_api/hot/list 纯 HTTP 热门列表。
package kuaishou

import (
	"fmt"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"multilive/internal/core"
	"multilive/internal/platform"
)

const (
	hotAPI     = "https://live.kuaishou.com/live_api/hot/list"
	playerBase = "https://astar.cc.cd/kuaishou/%s"
)

const (
	defaultPages = 50
	maxPages     = 50
	pageWorkers  = 5
)

var nameRe = regexp.MustCompile(`^[A-Z0-9]+$`)

type Platform struct{ platform.Base }

func (Platform) Name() string    { return "kuaishou" }
func (Platform) KeepStale() bool { return true }

func (Platform) FallbackURL(r core.Room) string { return fmt.Sprintf(playerBase, r.RID) }

func (Platform) Parse(line string) ([]core.Source, error) {
	t := strings.TrimSpace(line)
	if t == "" {
		return nil, nil
	}
	name := t
	pages := defaultPages
	if idx := strings.LastIndex(strings.Split(t, "/")[len(strings.Split(t, "/"))-1], ":"); idx >= 0 {
		last := t[strings.LastIndex(t, ":")+1:]
		if n, err := strconv.Atoi(last); err == nil {
			pages = n
			name = t[:strings.LastIndex(t, ":")]
		}
	}
	if i := strings.LastIndex(name, "/live/"); i >= 0 {
		name = name[i+len("/live/"):]
	}
	name = strings.TrimSuffix(strings.ToUpper(name), "/")
	name = strings.TrimSpace(name)
	if !nameRe.MatchString(name) {
		return nil, nil
	}
	if pages < 1 {
		pages = 1
	}
	if pages > maxPages {
		pages = maxPages
	}
	return []core.Source{{Platform: "kuaishou", Kind: "list", Target: name, Meta: pages}}, nil
}

func fetchPage(source string, page int) []map[string]any {
	sess := core.NewClient(nil)
	qs := url.Values{}
	qs.Set("type", source)
	qs.Set("filterType", "0")
	qs.Set("page", strconv.Itoa(page))
	qs.Set("pageSize", "24")
	var j map[string]any
	_, _, err := sess.GetJSON(hotAPI+"?"+qs.Encode(), "https://live.kuaishou.com/live/"+source, &j, 20*time.Second)
	if err != nil {
		core.Infof("  [%s] 第%d页失败: %s", source, page, core.FmtErr(err, 120))
		return nil
	}
	var out []map[string]any
	for _, it := range core.JList(core.JMap(j, "data"), "list") {
		if m, ok := it.(map[string]any); ok {
			out = append(out, m)
		}
	}
	return out
}

// hasPlayURL 列表自带 playUrls 是否有可用直链（仅作在播判断，不直接写直链）。
func hasPlayURL(room map[string]any) bool {
	type rep struct {
		level   float64
		bitrate float64
		url     string
	}
	var reps []rep
	if pus, ok := room["playUrls"].([]any); ok {
		for _, pu := range pus {
			pum, _ := pu.(map[string]any)
			if pum == nil {
				continue
			}
			as, _ := core.JMap(pum, "adaptationSet")["representation"].([]any)
			for _, r := range as {
				rm, _ := r.(map[string]any)
				if rm == nil {
					continue
				}
				u, _ := rm["url"].(string)
				if u == "" {
					continue
				}
				lv, _ := rm["level"].(float64)
				br, _ := rm["bitrate"].(float64)
				reps = append(reps, rep{lv, br, u})
			}
		}
	}
	if len(reps) == 0 {
		return false
	}
	sort.Slice(reps, func(i, j int) bool {
		if reps[i].level != reps[j].level {
			return reps[i].level < reps[j].level
		}
		return reps[i].bitrate < reps[j].bitrate
	})
	return reps[len(reps)-1].url != ""
}

func fetchSource(source string, pages int) map[string]map[string]any {
	rooms := map[string]map[string]any{}
	var mu sync.Mutex
	nums := make([]int, pages)
	for i := range nums {
		nums[i] = i + 1
	}
	results := core.Parallel(pageWorkers, nums, func(p int) []map[string]any {
		return fetchPage(source, p)
	})
	for _, list := range results {
		for _, room := range list {
			lid := core.JStr(room, "id")
			if lid == "" {
				continue
			}
			mu.Lock()
			if _, ok := rooms[lid]; !ok {
				rooms[lid] = room
			}
			mu.Unlock()
		}
	}
	core.Infof("  [%s] %d页抓完: %d 个不重复房间", source, pages, len(rooms))
	return rooms
}

func (Platform) Fetch(sources []core.Source, ctx *core.Ctx) ([]core.Room, error) {
	core.Infof("[kuaishou] %d 个来源", len(sources))
	type req struct {
		target string
		pages  int
	}
	var reqs []req
	for _, s := range sources {
		pages := s.Meta
		if pages <= 0 {
			pages = defaultPages
		}
		if pages > maxPages {
			pages = maxPages
		}
		if ctx.PagesCap > 0 && pages > ctx.PagesCap {
			pages = ctx.PagesCap
		}
		reqs = append(reqs, req{s.Target, pages})
	}
	results := core.Parallel(5, reqs, func(r req) map[string]map[string]any {
		return fetchSource(r.target, r.pages)
	})
	all := map[string]map[string]any{}
	for _, m := range results {
		for lid, room := range m {
			if _, ok := all[lid]; !ok {
				all[lid] = room
			}
		}
	}
	var out []core.Room
	seen := map[string]bool{}
	for _, room := range all {
		author := core.JMap(room, "author")
		rid := core.JStr(author, "id")
		if rid == "" || seen[rid] {
			continue
		}
		seen[rid] = true
		if !hasPlayURL(room) {
			continue
		}
		game := strings.TrimSpace(core.JStr(core.JMap(room, "gameInfo"), "name"))
		if game == "" {
			game = "kuaishou"
		}
		out = append(out, core.Room{
			Platform: "kuaishou", RID: rid,
			Title:    strings.TrimSpace(core.JStr(room, "caption")),
			Nickname: strings.TrimSpace(core.JStr(author, "name")),
			URL:      fmt.Sprintf(playerBase, rid),
			Group:    game, Avatar: core.JStr(room, "cover"),
		})
	}
	core.Infof("[kuaishou] 完成: %d 房间（无直链/无 author.id 已跳过）", len(out))
	return out, nil
}

var _ platform.Platform = Platform{}
